### PR TITLE
Implement fromChunksIO and update docs

### PR DESCRIPTION
**Pull request by adithyaov**
Thanks to @hasufell for the notification of the issue :-)

Refer #7

### CODE REVIEW

**Comment by adithyaov** [comment]
@hasufell It would be great if you could find time to review this PR :-)

File: src/Streamly/External/ByteString/Lazy.hs [hunk 1]
```diff
@@ -43,7 +48,55 @@ read = concat readChunks A.read
...
+--
+-- We can compose /liftToLazy/ with /liftLazyToM/ (/liftLazyToM . liftToLazy/) to
+-- lift the evaluation to the desired monad. Please note that the desired monad
+-- here should be lazy in nature.
 {-# INLINE fromChunks #-}
 fromChunks :: Monad m => SerialT m (Array Word8) -> m ByteString
 fromChunks = S.foldr BSLI.chunk Empty . S.map Strict.fromArray
+
+-- | Convert a serial stream of 'Array' 'Word8' to a lazy 'ByteString' in the /IO/
+-- monad.
```
Comment (hasufell) [inline src/Streamly/External/ByteString/Lazy.hs]: I think we should explicitly mention that all IO here will be evaluated lazily.

File: README.md [hunk 2]
```diff
@@ -1,35 +1,103 @@
...
 Library for streamly and bytestring interoperation.
 
+`Streamly.Memory.Array` is a generalization of `ByteString`.
+`Streamly.Memory.Array` used proper monadic programming instead
+of `unsafePerformIO`.
+
+This library is for interoperation with legacy programs that
+already use `ByteString`. In case you are writing code from scratch,
+consider using `Streamly.Memory.Array`.
+
```
Comment (harendra-kumar) [inline README.md:10]: If you are writing code from scratch, please use `Streamly.Memory.Array` which is a generalization of `ByteString` and better integrated with streamly. This library is to enable interoperation of streamly with existing code that uses `ByteString`.  

The package provides APIs to interconvert between strict Bytestring and streamly `Array Word8`  and between lazy `Bytestring` and stream of `Array Word8` . The interconversion has no overhead as the underlying representation of ByteString and Array are the same, we just need to rewrap the data in a different type.

File: README.md [hunk 3]
```diff
@@ -1,35 +1,103 @@
...
 ## Description
```
Comment (harendra-kumar) [inline README.md]: This section can be removed. This is mostly reference documentation which should be in the haddock docs.

File: src/Streamly/External/ByteString/Lazy.hs [hunk 4]
```diff
@@ -43,7 +48,61 @@ read = concat readChunks A.read
...
+-- /fromChunks/ can then be used as,
+-- @
+-- {-# INLINE fromChunksIO #-}
+-- fromChunksIO :: SerialT IO (Array Word8) -> IO ByteString
+-- fromChunksIO str = runLazy (fromChunks (S.hoist liftToLazy str))
+-- @
+--
+-- We can compose /liftToLazy/ with /liftLazyToM/ (/liftLazyToM . liftToLazy/) to
+-- lift the evaluation to the desired monad. Please note that the desired monad
+-- here should be lazy in nature.
```
Comment (harendra-kumar) [inline src/Streamly/External/ByteString/Lazy.hs]: We can rewrite line 64-70 like this:

IMPORTANT NOTE: This function is lazy only for lazy monads (e.g. Identity). For strict monads (e.g. IO) it consumes the entire input before generating the output. For IO monad please use `fromChunksIO` instead.

For strict monads like IO you could create a `newtype` wrapper to make the monad bind operation lazy and lift the stream to that type using `hoist`, then you can use this function to generate the bytestring lazily. For example you can wrap the IO type to make the bind lazy like this:

File: src/Streamly/External/ByteString/Lazy.hs [hunk 5]
```diff
@@ -43,7 +48,61 @@ read = concat readChunks A.read
...
 {-# INLINE fromChunks #-}
 fromChunks :: Monad m => SerialT m (Array Word8) -> m ByteString
 fromChunks = S.foldr BSLI.chunk Empty . S.map Strict.fromArray
+
+-- | Convert a serial stream of 'Array' 'Word8' to a lazy 'ByteString' in the /IO/
+-- monad. Although the /IO/ monad is strict in nature we emulate laziness
+-- using 'unsafeInterleaveIO'.
+--
+-- You must almost always use this function over 'fromChunks' if the computation
+-- is in the /IO/ monad.
```
Comment (harendra-kumar) [inline src/Streamly/External/ByteString/Lazy.hs]: The text from "Although" onwards can be removed.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Comment by adithyaov** [comment]
@psibi Can you please make another release when you find the time. Thanks!

**Pull request merged**